### ISSUE TITLE
local ip address support instead of forcing "localhost"

### ISSUE THREAD

**Issue by ejkkan** [nhost/nhost#1550]
**Is your feature request related to a problem? Please describe.**
When using android emulators for react native the norm is to use the networks ip since "localhost" is appointed to the emulator's own localhost.

**Describe the solution you'd like**
I would like to be able to set my network ip instead of enforcing and shutting down when not using localhost.
Example: 
`
const nhost = new NhostClient({
  subdomain: '192.168....',
  region: '',
});
`
Currently nhost throws: 
`
Error: "region" must be set when using a "subdomain" other than "localhost".
`

**Comment by guicurcio** [comment]
This is rather a Typescript issue than something to implement in the SDK (we have a PR open) I think you should be able to do something like this:

```
import { NhostClient } from "@nhost/nextjs";
import { NhostClientConstructorParams } from "@nhost/nhost-js";

type NewNhostClient = Omit<
  NhostClientConstructorParams,
  | "clientStorage"
  | "clientStorageType"
  | "clientStorageGetter"
  | "clientStorageSetter"
>;

export const nhost = new NhostClient({
  authUrl: "https://hasura-auth.....",
  graphqlUrl: "http://localhost:8080/v1/graphql",
  storageUrl: "asdaasd",
  functionsUrl: "asdasd",
} as NewNhostClient);

export default nhost;
```

lmk if it works

**Comment by ejkkan** [comment]
Hey, thanks for getting back to me!

You're using the nextjs lib in your example, not sure if it's on-par with the regular react package.

I tried adding it `authUrl` to the `NhostClient` and it does not affect the behavior.
I've tested that everything is working fine on iOS when using 'localhost'. In theory it should work on android just by switching 'localhost' to the computer ip to reach the localhost. 

If I try to start the app with using the ip, the app won't start. So it seems to be more than just a TS warning in it's not compiling. 
I'm using 'http://192.168.100.147:1337/v1/auth/signin/provider/google' to reach the local auth sessions which at least takes me to the social login screen. But the return url takes me to 'localhost' which won't work.

I'll attach a video of what happens.



https://user-images.githubusercontent.com/32518962/216848104-07bff367-4a61-4214-910c-9f2ea73f834a.mp4

**Comment by elitan** [comment]
Was this fixed in relation to this issue: https://github.com/nhost/nhost/issues/1551 ?

**Comment by ejkkan** [comment]
Hey @elitan.
Unfortunately no, this is a separate issue.

**Comment by rvalledorjr** [comment]
up for this. In my case, my Nuxt app is running inside a docker container. And during server-side rendering it makes some nhost call determining the currently logged-in user. The call always fails with an ECONNREFUSED error, which I think is due to `localhost` resolving to its ipv6 address `::1`. I tried changing the subdomain value from `localhost` to its ipv4 address `127.0.0.1` but I got the same error @ejkkan had.

**Comment by pkreipke** [comment]
@rvalledorjr Does this help: using host.docker.internal.... ?

https://github.com/nhost/cli/issues/501

**Comment by ejkkan** [comment]
Unfortunately not.
Using anything else then 'localhost' throws an error with 'region' must be used.
I'm not sure there's any other solution than to accept other string than localhost for 'subdomain'

**Comment by szilarddoro** [comment]
@ejkkan do you have a reproduction repository where I can check the problem?

**Comment by ejkkan** [comment]
@szilarddoro sorry for the late reply.
I do not have one. But an easy way to set it up would be to setup a new expo app.
Easiest is either via [expo itself](https://docs.expo.dev/tutorial/create-your-first-app/) or a [boilerplate](https://github.com/obytes/react-native-template-obytes). I'm using the boilerplate for my app for example.

After that just setup Nhost like you would any web project and run it on an android device.

**Comment by szilarddoro** [comment]
Could you please give [custom URLs](https://docs.nhost.io/reference/javascript#using-custom-urls) a try?

**Comment by pchmn** [comment]
I had the same problem, so I ended creating a little proxy server that I start before starting my react-native app. The proxy server listens on port `5050` and redirects to the desire nhost service. For example to call the auth service, on another machine on my network, I can call `http://<local-ip-adress>:5050/auth` and it redirects to `https://local.auth.nhost.run/v1`. 

My proxy server looks like that:
```ts
import { createServer } from 'http';
import { createProxyServer } from 'http-proxy';

const proxy = createProxyServer({ changeOrigin: true });

const nHostUrls = {
  wsGraphql: 'wss://local.hasura.nhost.run/v1',
  graphql: 'https://local.graphql.nhost.run/v1',
  auth: 'https://local.auth.nhost.run/v1',
  storage: 'https://local.storage.nhost.run/v1',
  functions: 'https://local.functions.nhost.run/v1',
  mailhog: 'https://local.mailhog.nhost.run/v1',
  // Access to dashboards
  hasura: 'https://local.hasura.nhost.run',
  dashboard: 'https://local.dashboard.nhost.run',
};

// Http proxy
const server = createServer((req, res) => {
  const nhostService = req.url?.split('/')[1];

  if (!nhostService) {
    res.writeHead(404);
    res.end('Not found');
    return;
  }

  req.url = req.url?.replace(`/${nhostService}`, '') || '';

  const target = nHostUrls[nhostService as keyof typeof nHostUrls];

  proxy.web(req, res, { target });
});

// Websocket proxy
server.on('upgrade', (req, socket, head) => {
  proxy.ws(req, socket, head, { target: nHostUrls.wsGraphql });
});

server.listen(5050);
console.log('listening on port 5050');
```

And I use this configuration in my react native app: 
```tsx
const nhost = new NhostClient({
  // local ip adress in the form 192.168.x.x
  authUrl: 'http://<local-ip-adress>:5050/auth',
  graphqlUrl: 'http://<local-ip-adress>:5050/graphql',
  storageUrl: 'http://<local-ip-adress>:5050/storage',
  functionsUrl: 'http://<local-ip-adress>:5050/functions',
});
```

**Comment by dbarrosop** [comment]
an easier solution would be to just add to `/etc/hosts` (or equivalent in expo) the URLs pointing to the machine where you are running Nhost, for instance:


```
192.68.1.150 local.auth.nhost.run
192.68.1.150 local.db.nhost.run
192.68.1.150 local.functions.nhost.run
192.68.1.150 local.graphql.nhost.run
192.68.1.150 local.hasura.nhost.run
192.68.1.150 local.storage.nhost.run
```

**Comment by pchmn** [comment]
It works without configuration on an iOS simulator because it is on the same network as my mac. I didn't test it on Android simulator, but I think you'd have to modify simulator `system/hosts` as you say.

But when using a physical device, on Android you can't modify `system/hosts` if you're not root, and on iOS I don't think you can modify it.

This is why I created this proxy server

**Comment by dbarrosop** [comment]
ok, that makes sense and it is a great solution. Maybe a similar solution is something that can be added to the CLI. It shouldn't be too difficult to add the necessary changes to the cli (traefik) to allow accepting something like this, the only issue is that then you need to specify all urls like you did.

**Comment by pchmn** [comment]
After looking at how urls are constructed in `NhostClient` ([here](https://github.com/nhost/nhost/blob/main/packages/nhost-js/src/utils/helpers.ts#L15)), we don't need to specify all urls, but just the `backendUrl`. I just had to modify minor things in my proxy (remove `/v1` in my target urls and get second element of `req.url?.split('/')` to retrieve nhost service):

```typescript
import { createServer } from 'http';
import { createProxyServer } from 'http-proxy';

const proxy = createProxyServer({ changeOrigin: true });

const nHostUrls = {
  wsGraphql: 'wss://local.hasura.nhost.run',
  graphql: 'https://local.graphql.nhost.run',
  auth: 'https://local.auth.nhost.run',
  storage: 'https://local.storage.nhost.run',
  functions: 'https://local.functions.nhost.run',
};

// Http proxy
const server = createServer((req, res) => {
  // When using backendUrl, req.url is in the form '/v1/<service>/<route>'
  const nhostService = req.url?.split('/')[2];

  if (!nhostService) {
    res.writeHead(404);
    res.end('Not found');
    return;
  }

  req.url = req.url?.replace(`/${nhostService}`, '') || '';

  const target = nHostUrls[nhostService as keyof typeof nHostUrls];

  proxy.web(req, res, { target });
});

// Websocket proxy
server.on('upgrade', (req, socket, head) => {
  proxy.ws(req, socket, head, { target: nHostUrls.wsGraphql });
});

server.listen(5050);
console.log('listening on port 5050');
```

And then in react-native app, just use `backendUrl` instead of all urls:
```typescript
const nhost = new NhostClient({
  backendUrl: 'http://<local-ip-address>:5050',
});
```

**Comment by dbarrosop** [comment]
backendURL is indeed deprecated so, right now, specifying all URLs like you did is the correct approach.

**Comment by rsnider-wisecode** [comment]
@pchmn thanks for the proxy code here, it's working to get the auth flow started. how did you handle auth redirects with this? when i do google login, my phone is still redirected to local.auth.nhost.run. i assume it is because the docker contianer config is specifying that as the domain

**Comment by dbarrosop** [comment]
I'd suggest taking a look at this document with our recommended solution for this type of environment:

https://docs.nhost.io/guides/cli/connect-devices-to-local-nhost-project

**Comment by rsnider-wisecode** [comment]
Oh nice, thanks! I hadn't found that yet

**Comment by dbarrosop** [comment]
We are planning to release a better solution very soon that won't require any external services or configuration on the emulator/device (next week most likely) so I'd suggest you to stay tuned (we usually post things like this in our discord and/or our monthly releases in the nhost/nhost repo).

**Comment by dbarrosop** [comment]
This should solve this issue in an easy way without requiring proxies, dnsmasq or anything extra:

https://docs.nhost.io/guides/cli/subdomain
